### PR TITLE
Add filter pushdown API for contains and endsWith

### PR DESCRIPTION
**Pull request by yabola**
This PR adds filter pushing down api for `endsWith` (`like %x`) and `contains` (`like %x%`). Before this PR, iceberg only support pushdown startWith filter. Spark parquet supports pushdown `endWiths` and `contains`.
This PR is the first part, the next PR will complete the related implementation of spark and fileFormat

### CODE REVIEW

**Comment by yabola** [comment]
The origin PR is https://github.com/apache/iceberg/pull/9683 , I split it into two PRs and add some UT.

**Comment by yabola** [comment]
@amogh-jahagirdar Hi, if you have time, please take a look~

**Comment by amogh-jahagirdar** [comment]
Sorry for the delay @yabola , got busy with some other work, I haven't forgotten about this PR! When my time frees up I'll review this.

File: api/src/test/java/org/apache/iceberg/expressions/TestEvaluator.java [hunk 1]
```diff
@@ -348,6 +352,102 @@ public void testNotStartsWith() {
...
+    assertThat(evaluator.eval(TestHelpers.Row.of("aabca")))
+        .as("aabca contains abc should be true")
+        .isTrue();
+    assertThat(evaluator.eval(TestHelpers.Row.of((String) null)))
+        .as("null contains abc should be false")
+        .isFalse();
+  }
+
+  @Test
+  public void testNotContains() {
```
Comment (amogh-jahagirdar) [inline api/src/test/java/org/apache/iceberg/expressions/TestEvaluator.java:428]: Can you add a notContains test for a null row?

File: api/src/main/java/org/apache/iceberg/expressions/ExpressionVisitors.java [hunk 2]
```diff
@@ -266,6 +294,22 @@ public <T> R notStartsWith(Bound<T> expr, Literal<T> lit) {
       throw new UnsupportedOperationException("Unsupported operation.");
     }
 
+    public <T> R endsWith(Bound<T> expr, Literal<T> lit) {
+      throw new UnsupportedOperationException("Unsupported operation.");
```
Comment (amogh-jahagirdar) [inline api/src/main/java/org/apache/iceberg/expressions/ExpressionVisitors.java:298]: Could you include the relevant details in these exception messages? "endsWith is an unsupported operation" etc.

File: api/src/test/java/org/apache/iceberg/expressions/TestEvaluator.java [hunk 3]
```diff
@@ -348,6 +352,102 @@ public void testNotStartsWith() {
...
+        .isTrue();
+    assertThat(evaluator.eval(TestHelpers.Row.of("c")))
+        .as("c notEndsWith abc should be true")
+        .isTrue();
+    assertThat(evaluator.eval(TestHelpers.Row.of("ababc")))
+        .as("ababc notEndsWith abc should be false")
+        .isFalse();
+    assertThat(evaluator.eval(TestHelpers.Row.of("AbAbc")))
+        .as("AbAbc notEndsWith abc should be true")
+        .isTrue();
```
Comment (amogh-jahagirdar) [inline api/src/test/java/org/apache/iceberg/expressions/TestEvaluator.java:400]: Same here could you add an`notEndsWith" test for null? Also testing empty string would be nice.

File: api/src/test/java/org/apache/iceberg/transforms/TransformTestHelpers.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.Literal;
+import org.apache.iceberg.expressions.Projections;
+import org.apache.iceberg.expressions.UnboundPredicate;
+import org.apache.iceberg.types.Types;
+
+public class TransformTestHelpers {
```
Comment (amogh-jahagirdar) [inline api/src/test/java/org/apache/iceberg/transforms/TransformTestHelpers.java:31]: Instead of a helper class, could we make this a base class for all the transform tests; `TestTransformBase` or something like that? Then TestContains, TestEndsWith  can all inherit that class and use the protected assertProjection methods.

File: api/src/test/java/org/apache/iceberg/transforms/TestNotEndsWith.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        spec, notEndsWith(COLUMN, "ab"), "ab", Expression.Operation.NOT_ENDS_WITH);
+    TransformTestHelpers.assertProjectionStrict(
+        spec, notEndsWith(COLUMN, "abab"), "abab", Expression.Operation.NOT_EQ);
+    TransformTestHelpers.assertProjectionStrict(
+        spec, notEndsWith(COLUMN, "abcdab"), "abcd", Expression.Operation.NOT_ENDS_WITH);
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void testEvaluatorTruncateString() {
```
Comment (amogh-jahagirdar) [inline api/src/test/java/org/apache/iceberg/transforms/TestNotEndsWith.java:66]: I think we need more tests on different string length combinations and metrics evaluator. I think you can refer to `TestNotStartsWith` for good tests.

**Review by amogh-jahagirdar — CHANGES_REQUESTED** [review]
Thanks @yabola I think the logic looks good but I think there are more test cases we need to add to the transform tests.

File: api/src/test/java/org/apache/iceberg/expressions/TestEvaluator.java (hunk 1 above)
Comment (yabola) [reply api/src/test/java/org/apache/iceberg/expressions/TestEvaluator.java:458]: add `new Object[] {null}` and empty string

**Comment by yabola** [comment]
@amogh-jahagirdar Thanks for your review! I have updated my code.

File: api/src/test/java/org/apache/iceberg/transforms/TestNotStartsWith.java [hunk 6]
```diff
@@ -106,23 +75,23 @@ public void testTruncateStringWhenProjectedPredicateTermIsLongerThanWidth() {
         .isEqualTo("ab");
 
     assertThat(evaluator.eval(TestHelpers.Row.of("abcde")))
-        .as("notStartsWith(abcde, truncate(abcde,2)) => false")
+        .as("notStartsWith('abcde', truncate(abcde,2)) => false")
         .isFalse();
 
     assertThat(evaluator.eval(TestHelpers.Row.of("ab")))
-        .as("notStartsWith(abcde, truncate(ab, 2)) => false")
+        .as("notStartsWith('ab', truncate('abcde', 2)) => false")
```
Comment (yabola) [inline api/src/test/java/org/apache/iceberg/transforms/TestNotStartsWith.java:82]: I think the description before is unclear. I have modified it to a string value description.

**Comment by yabola** [comment]
@amogh-jahagirdar Hi, please take a look if you have time~

**Comment by yabola** [comment]
@sujithjay could you also take a look since you implements `startWith`~
Spark on parquet also support `endsWith` and `contains`
https://github.com/apache/spark/blob/356830ada6c6ebbf54e7852c37266c32bfa137ea/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala#L816-L844
